milestones/api.py

```diff
@@ -0,0 +1,312 @@
+"""
+api.py is an interface module for Python-level integration with the
+Milestones app.
+
+In the current incarnation of this particular application, the API
+operations are equivalent to the orchestration layer, which manages
+the application's workflows.
+
+Helplful Hint:  When modeling Milestones, I've found that it's helpful
+to first consider the process for fulfilling the Milestone, and then
+consider the process for requiring the Milestone.
+
+Note the terminology difference at this layer vs. Data -- add/edit/get/remove
+"""
+from . import data
+from . import exceptions
+from . import validators
+
+
+# PRIVATE/INTERNAL FUNCTIONS
+
+def _validate_course_key(course_key):
+    """ Validation helper """
+    if not validators.course_key_is_valid(course_key):
+        exceptions.raise_exception(
+            "CourseKey",
+            course_key,
+            exceptions.InvalidCourseKeyException
+        )
+
+
+def _validate_content_key(content_key):
+    """ Validation helper """
+    if not validators.content_key_is_valid(content_key):
+        exceptions.raise_exception(
+            "ContentKey",
+            content_key,
+            exceptions.InvalidContentKeyException
+        )
+
+
+def _validate_milestone(milestone):
+    """ Validation helper """
+    if not validators.milestone_is_valid(milestone):
+        exceptions.raise_exception(
+            "Milestone",
+            milestone,
+            exceptions.InvalidMilestoneException
+        )
+
+
+def _validate_milestone_relationship_type(name):
+    """ Validation helper """
+    if not validators.milestone_relationship_type_is_valid(name):
+        exceptions.raise_exception(
+            "MilestoneRelationshipType",
+            name,
+            exceptions.InvalidMilestoneRelationshipTypeException
+        )
+
+
+def _validate_user(user):
+    """ Validation helper """
+    if not validators.user_is_valid(user):
+        exceptions.raise_exception(
+            "User",
+            user,
+            exceptions.InvalidUserException
+        )
+
+
+# PUBLIC FUNCTIONS
+def add_milestone(milestone):
+    """
+    Passes a new milestone to the data layer for storage
+    """
+    _validate_milestone(milestone)
+    milestone = data.create_milestone(milestone)
+    return milestone
+
+
+def edit_milestone(milestone):
+    """
+    Passes an updated milestone to the data layer for storage
+    """
+    _validate_milestone(milestone)
+    try:
+        return data.update_milestone(milestone)
+    except exceptions.InvalidMilestoneException:
+        raise
+
+
+def get_milestone(milestone_id):
+    """
+    Retrieves the specified milestone
+    """
+    milestone = {
+        'id': milestone_id,
+    }
+    milestones = data.fetch_milestones(milestone)
+    if not len(milestones):
+        return None
+    return milestones[0]
```

Comment:
Given the ID is unique, it seems logical to have a get_milestone() function exposed from the data layer.  I think maybe I'm missing something here. I thought IntegrityErrors tended to occur in our DB due to write operations, or a "get_or_create" operation. A simple "get" should be safe, right? 

Also, I wonder if it's worth having a UUID for milestones, rather than passing around the PK of the table.